### DISCUSSION TITLE
Api route can't serve files after upgrading Ubuntu and docker

### DISCUSSION THREAD

**Discussion by github0013** [vercel/next.js#47590] [Help] [answered]
### Summary

I am not sure where to ask.  This can be related to docker / OS itself.   I am hoping someone had similar issues in the past.

I have 2 servers.  Ubuntu 20 and Ubuntu 22.   Since I installed Ubuntu 22 later, docker and docker compose are newer versions (as the descriptions below).

When I do run the EXACT same `/pages/api/pdf.ts` file, on Ubuntu 20, I get the PDF on my browser.  However on Ubuntu 22, I get empty JSON data `{}` coming back from the server.

<img width="672" alt="Screenshot 2023-03-28 at 14 28 33" src="https://user-images.githubusercontent.com/809378/228137117-73b280b8-f08b-4346-9366-35b81d0a2591.png">

Since I am using docker, it's based on the same image, same package versions, and same everything, except the docker and OS.

Can anybody take a guess where to look to find the cause?

### client
```
OSX 13.2.1 (22D68)

Google chrome
Version 111.0.5563.64 (Official Build) (arm64)
```

### Dockerfile
```
FROM node:lts-alpine
ENV LANG C.UTF-8
RUN apk update && apk upgrade
WORKDIR /app
```

### package.json
```
{
  "scripts": {
    "dev": "next",
  },
  "dependencies": {
    "next": "^13.2.4",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@types/node": "18.15.10",
    "@types/react": "18.0.30",
    "typescript": "5.0.2"
  }
}
```

### /pages/api/pdf.ts
```ts
import type { NextApiRequest, NextApiResponse } from "next"

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    return fetch(
      "https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf"
    ).then((response) => {
      res.setHeader("Content-Type", "application/pdf")
      res.send(response.body)
    })
  } catch (e) {
    console.log("e:", e)
    res.status(404).end()
  }
}
```

## Environment that works (A)
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.4 LTS
Release:        20.04
Codename:       focal

$ docker -v
Docker version 20.10.17, build 100c701

$ docker-compose -v
docker-compose version 1.29.1, build c34c88b2
```

## Environment that doesn't work (B)
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy

$ docker -v
Docker version 23.0.1, build a5ee5b1

$ docker compose version
Docker Compose version v2.16.0
```

# When it works (env A)
<img width="1437" alt="Screenshot 2023-03-28 at 14 30 08" src="https://user-images.githubusercontent.com/809378/228137349-84fdedb4-76c9-4234-b72d-db3221b0628a.png">

# When it doesn't work (env B)
<img width="1404" alt="Screenshot 2023-03-28 at 14 12 31" src="https://user-images.githubusercontent.com/809378/228135817-9390bf3f-5b01-40e8-944e-08f38886078b.png">

**Comment by icyJoseph** [comment, marked as answer]
What's the Node version in both?

## Edit

The introduction of native fetch, might lead to problems,  when used to get data, server side, from 3rd parties.

**Reply by github0013** [reply]
oh.. of course, node matters...

## Environment that works (A)
```
$ node -v
v14.17.0
```

## Environment that doesn't work (B)
```
$ node -v
v18.15.0
```

This has to be the cause.   As I was using `FROM node:lts-alpine` in the Dockerfiles, the version is different every time the image is created...

So I digged deep to see what version worked on my case.   And I checked v18s, and they had the same issue.  Then I used `v17.9.1` that is one version before `v18.0.0`, and it worked!

https://nodejs.org/download/release/

I don't think this is a bug as v18 is already up to v18.15.0 at this point.    Does anyone know what is causing the empty JSON `{}` return instead of the PDF binary data?

**Reply by icyJoseph** [reply]
The problem might be that when Next.js detects global fetch, it doesn't polyfill it. And what do you know, Node 18 started to support native fetch, https://nodejs.org/en/blog/announcements/v18-release-announce#fetch-experimental.

Normally, Next.js would use `node-fetch` to polyfill, but as said before, if it detects a global fetch, it takes that.

However, the native fetch is closer to the standard definition, so it doesn't make assumptions/fallbacks made by other fetch polyfills, the likes of node-fetch, whatwg-fetch, so there might be divergence in behavior. I'll try to find a link I had listing the how the native fetch is "different" than the others.

## Edit

Making the assumption here that the issue is with fetching the PDF server side and then sending it to the client. 

However, there might be other issues with the underlying HTTP handling of Node.js - ¯\\_(ツ)_/¯

**Reply by icyJoseph** [reply]
And please don't forget to mark as answered :) 😅
